src/compiler/program.ts

```diff
@@ -4357,6 +4357,15 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
             }
         }
 
+        if (options.noCheck) {
+            if (options.noEmit) {
+                createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_with_option_1, "noCheck", "noEmit");
+            }
+            if (!options.emitDeclarationOnly) {
+                createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_without_specifying_option_1, "noCheck", "emitDeclarationOnly");
+            }
+        }

```

Comment:
for now we should disable this for incremental as the build info generated will not handle this correctly.

Reply:
As in I should add an error for using `noCheck` alongside `incremental`?

Reply:
> or we need to make sure noCheck recalculates semantic diagnostics and stores the that flag into buildinfo

That's just setting `affectsSemanticDiagnostics` in the option descriptor when we add it back to being enabled for cli/tsconfig in the future, right?

Reply:
"in option" yes but now it needs to be handled in builder or reported as error to have incremental and nocheck as buildinfo written will not be correct. It will show no errors and next time you run tsc it will not report errors . (without storing noCheck value in buildInfo and checking for that in builder)

Reply:
Is [051a1d1](https://github.com/microsoft/TypeScript/pull/57934/commits/051a1d16e87bb418e44d1dd56a1d689cd3982fe0) along the lines of what you were thinking? Since it's not in the `commonOptionsWithBuild`, `commandOptionsWithoutBuild`, or `optionDeclarations` lists, it's never included in `help` or valid on the CLI/in a config, but since it's built into the `OptionsNameMap`, all the normal option serializing/deserializing machinery works with it - build info included.

Reply:
We need test cases: 

1. Run through API with `--no-check`  , followed by `tsc` .(with and without incremental). 
2.  Run through API with `--no-check` , followed by `tsc --b` (with and without incremental) - have a feeling that tsc - in non incremental is going to find that everything is uptodate (there is no buildinfo to check, only output file stamps) and will not report errors? 

You can add these testcases by using `verifyTsc` where the initial `fs` is generated through running API with `--no-check`

Reply:
> Run through API with --no-check

We.... don't have the harness for this right now. At least not a harness that then feeds into the existing `verifyTsc` stuff (obviously, all the `compiler`-style tests build against the API without using the `executeCommandLine` entrypoint). This is quite a lot to put together for something that's only supposed to be API-only temporarily... I started putting it together, and such a harness is looking like it's going to be more than the entire rest of the PR. 😵

So I'm not going to. This is gonna be a normal command line option, accepted on both the CLI and via `tsconfig`, except with a
```ts
        extraValidation(value) {
            return [Diagnostics.Unknown_compiler_option_0, "noCheck"];
        },
```
so anything that runs validation (API methods that take `CompilerOptions` obviously do not) reports that it doesn't exist, but otherwise, it uses all the usual pipelines for things. And then in the test where I'd like to consider it a real option, I'll override that validator with a `noop` temporarily, which I can do for the `tsc -b` style tests, to validate it will work correctly with `-b` once it's actually public and usable via.... anything other than the API.